src/session_base.cpp

```diff
@@ -550,6 +550,54 @@ void zmq::session_base_t::reconnect ()
         _pipe->hiccup ();
 }
 
+zmq::session_base_t::connecter_factory_entry_t
+  zmq::session_base_t::_connecter_factories[] = {
+    connecter_factory_entry_t (protocol_name::tcp,
+                               &zmq::session_base_t::create_connecter_tcp),
+#if !defined ZMQ_HAVE_WINDOWS && !defined ZMQ_HAVE_OPENVMS                     \
+  && !defined ZMQ_HAVE_VXWORKS
+    connecter_factory_entry_t (protocol_name::ipc,
+                               &zmq::session_base_t::create_connecter_ipc),
+#endif
+#if defined ZMQ_HAVE_TIPC
+    connecter_factory_entry_t (protocol_name::tipc,
+                               &zmq::session_base_t::create_connecter_tipc),
+#endif
+#if defined ZMQ_HAVE_VMCI
+    connecter_factory_entry_t (protocol_name::vmci,
+                               &zmq::session_base_t::create_connecter_vmci),
+#endif
+};
+
+zmq::session_base_t::connecter_factory_map_t
+  zmq::session_base_t::_connecter_factories_map (
```

Comment:
@sigiesec address sanitizer is complaining about uninitialized value here (line numbers have changed), any idea?

https://travis-ci.org/github/google/oss-fuzz/jobs/680325364

```
  Uninitialized value was created by a heap allocation
    #0 0x525d89 in operator new(unsigned long) /src/llvm-project/compiler-rt/lib/msan/msan_new_delete.cpp:45:35
    #1 0x741319 in std::__1::unique_ptr<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>, std::__1::__tree_node_destructor<std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*> > > > std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::__construct_node<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&>(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&) /usr/local/bin/../include/c++/v1/__tree:2190:23
    #2 0x7402c8 in std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long> std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::__emplace_hint_unique_key_args<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&>(std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&) /usr/local/bin/../include/c++/v1/__tree:2169:29
    #3 0x73fa3d in std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long> > std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool), std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::insert<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>&, void>(std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long> >, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>&) /usr/local/bin/../include/c++/v1/map:1167:29
    #4 0x73f46c in void std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool), std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::insert<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*>(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*) /usr/local/bin/../include/c++/v1/map:1199:17
    #5 0x73cfe1 in std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool), std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::map<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*>(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > const&) /usr/local/bin/../include/c++/v1/map:984:13
    #6 0x41f037 in __cxx_global_var_init.23 /src/libzmq/src/session_base.cpp:626:24
    #7 0x41f2a5 in _GLOBAL__sub_I_session_base.cpp /src/libzmq/src/session_base.cpp
    #8 0x8d0f6c in __libc_csu_init (/tmp/not-out/test_connect_null_fuzzer+0x8d0f6c)
SUMMARY: MemorySanitizer: use-of-uninitialized-value /usr/local/bin/../include/c++/v1/string:947:17 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::size() const
```

Reply:
Reverted this commit via https://github.com/zeromq/libzmq/pull/3896